### DISCUSSION TITLE
PKCS12_create in fips mode (openssl 3.4)

### DISCUSSION THREAD

**Discussion by ChauhanPriyanka** [openssl/openssl#25897] [Q&A] [answered]
Hi team,

I'm trying to validate the PKCS12_create api in fips mode, without fips the api in below sample is working fine.
Getting below error in FIPS mode, please let me know what I'm doing wrong.

> FIPS provider loaded
Base provider loaded
Private Key (cert.key) written successfully
Certificate (cert.pem) written successfully
Error creating PKCS12 structure
Error: error:0308010C:digital envelope routines::unsupported
Error: error:1180006B:PKCS12 routines::key gen error
Error: error:1180006D:PKCS12 routines::mac generation error

Sample code
```
void handleErrors() {
    unsigned long errCode;
    while ((errCode = ERR_get_error()) != 0) {
        char* errMsg = ERR_error_string(errCode, NULL);
        std::cerr << "Error: " << errMsg << std::endl;
    }
}

int test_pkcs12_create_api() {
    // Initialize OpenSSL
    OSSL_PROVIDER* fips;
    OSSL_PROVIDER* base;

    OSSL_PROVIDER_set_default_search_path(NULL, "C:\\UnitTestOpenSSL34\\Debug");

    if (!OSSL_LIB_CTX_load_config(NULL, "C:\\UnitTestOpenSSL34\\Debug\\openssl.cnf"))
    {
        ERR_print_errors_cb(err_print_line, NULL);
    }

    fips = OSSL_PROVIDER_load(NULL, "fips");
    if (fips == NULL) {
        ERR_print_errors_cb(err_print_line, NULL);
        std::cout << "Failed to load FIPS provider"<< std::endl;
        exit(EXIT_FAILURE);
    }
    std::cout << "FIPS provider loaded"<< std::endl;

    base = OSSL_PROVIDER_load(NULL, "base");
    if (base == NULL) {
        //OSSL_PROVIDER_unload(fips);
        std::cout << "Failed to load base provider" << std::endl;
        exit(EXIT_FAILURE);
    }
    std::cout << "Base provider loaded" << std::endl;

    // Create a new EVP_PKEY context (RSA 2048)
    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
    if (!pctx) {
        handleErrors();
    }

    // Initialize key generation
    if (EVP_PKEY_keygen_init(pctx) <= 0) {
        handleErrors();
    }

    // Set RSA key size (2048 bits)
    if (EVP_PKEY_CTX_set_rsa_keygen_bits(pctx, 2048) <= 0) {
        handleErrors();
    }

    // Generate the private key
    EVP_PKEY* pkey = NULL;
    if (EVP_PKEY_keygen(pctx, &pkey) <= 0) {
        handleErrors();
    }

    // Create the private key file (cert.key)
    BIO* key_bio = BIO_new_file("C:\\UnitTestOpenSSL34\\Debug\\cert.key", "w");
    if (PEM_write_bio_PrivateKey(key_bio, pkey, NULL, NULL, 0, NULL, NULL) != 1) {
        handleErrors();
    }
    std::cout << "Private Key (cert.key) written successfully" << std::endl;

    // Create a self-signed certificate (X.509)
    X509* x509 = X509_new();
    if (!x509) {
        handleErrors();
    }

    // Set the version (X509 version 3)
    if (X509_set_version(x509, 2) != 1) {
        handleErrors();
    }

    // Set the serial number (use current timestamp for simplicity)
    ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);  // Simple serial number (1)

    // Set the issuer and subject to "self" (self-signed certificate)
    X509_NAME* name = X509_get_subject_name(x509);
    X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char*)"US", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "ST", MBSTRING_ASC, (unsigned char*)"Some State", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "L", MBSTRING_ASC, (unsigned char*)"Some City", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char*)"My Company", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char*)"localhost", -1, -1, 0);

    // Set the issuer as the subject (self-signed)
    if (X509_set_issuer_name(x509, name) != 1) {
        handleErrors();
    }

    X509_set_pubkey(x509, pkey);

    // Set the validity period (e.g., 365 days)
    ASN1_TIME* notBefore = ASN1_TIME_set(NULL, time(NULL));
    ASN1_TIME* notAfter = ASN1_TIME_set(NULL, time(NULL) + 365 * 24 * 60 * 60); // 1 year validity

    if (X509_set_notBefore(x509, notBefore) != 1) {
        handleErrors();
    }
    if (X509_set_notAfter(x509, notAfter) != 1) {
        handleErrors();
    }

    // Sign the certificate with the private key (self-signed)
    if (X509_sign(x509, pkey, EVP_sha256()) == 0) {
        handleErrors();
    }

    // Write the certificate to a file (cert.pem)
    BIO* cert_bio = BIO_new_file("C:\\UnitTestOpenSSL34\\Debug\\cert.pem", "w");
    if (PEM_write_bio_X509(cert_bio, x509) != 1) {
        handleErrors();
    }
    std::cout << "Certificate (cert.pem) written successfully" << std::endl;
    if (X509_check_private_key(x509, pkey))
    {
        STACK_OF(X509)* caCerts = nullptr;
        // Create the PKCS12 structure
        const char* password = ""; // The password to protect the PKCS12 file
        PKCS12* p12 = PKCS12_create("", "Friendly name", pkey, x509, nullptr, 0, 0, 0, 0, 0);

        if (!p12) {
            std::cerr << "Error creating PKCS12 structure" << std::endl;
            handleErrors();
            return -1;
        }

        // Save the PKCS12 structure to a file
        const char* p12File = "C:\\UnitTestOpenSSL34\\Debug\\output.p12";  // Output file
        BIO* p12BIO = BIO_new_file(p12File, "wb");
        if (!p12BIO) {
            perror("Error opening key file");
            BIO_free(p12BIO);
            return -1;
        }

        if (i2d_PKCS12_bio(p12BIO, p12) == 0) {
            std::cerr << "Error writing PKCS12 file" << std::endl;
            handleErrors();
            PKCS12_free(p12);
            return -1;
        }
        
        std::cout << "PKCS12 file created successfully: " << p12File << std::endl;

        PKCS12_free(p12);
        BIO_free(p12BIO);
        BIO_free_all(key_bio);
        BIO_free_all(cert_bio);
        X509_free(x509);
        EVP_PKEY_free(pkey);
    }
    return 0;
}
```

**Comment by beldmit** [comment, marked as answer]
I think you should look at https://github.com/openssl/openssl/blob/e54526413d5ef7c665e25f552f2f01d4352bd33d/apps/pkcs12.c#L694-L726

**Comment by ChauhanPriyanka** [comment]
Thank you for sharing the code snippet for reference @beldmit .
So I have removed the mac verification during pkcs12_create by setting maciter = -1 and doing mac verification after pkcs12 structure is created. Could you please help me to understand this.

1. The code for Mac construction using `PKCS12_set_pbmac1_pbkdf2` should be guarded with check - if fips provider is loaded.
2. When to use PKCS12_set_mac.

```
PKCS12* p12 = PKCS12_create("password", "Friendly name", pkey, x509, nullptr, NID_aes_256_cbc, NID_aes_256_cbc, 0, -1, 0);

if (!p12) {
    std::cerr << "Error creating PKCS12 structure" << std::endl;
    handleErrors();
    return -1;
}

// Password for key derivation
const char* password = "strongpassword";
int passlen = strlen(password);         // Length of the password
int iter = 1000;
const EVP_MD* md_type = EVP_sha256();    // SHA-256 as the hash function
const char* prf_md_name = "sha256";        

// Set PBKDF2 and PBMAC1 in the PKCS#12 structure
if (!PKCS12_set_pbmac1_pbkdf2(p12, password, passlen, NULL, 0, 1, md_type, prf_md_name)) {
    handleErrors();
}
```

**Reply by beldmit** [reply]
1. It shouldn't - you can create and verify the PBKDF2-based MAC both in FIPS and non-FIPS mode. On verification, the proper functions will be used based on MAC OID, see RFC 9579. You can't use the "classic" PKCS12 MAC in FIPS mode.
2. When you need to be interoperable with implementations that don't support RFC 9579.

**Comment by ChauhanPriyanka** [comment]
Hi @beldmit , 
I have created an RSA key of 2048 bits while executing the !`PKCS12_set_pbmac1_pbkdf2(p12, password, passlen, NULL, 0, 1, md_type, prf_md_name) ` I'm getting following error

> Error: error:02000078:rsa routines::key size too small. 

Please help what is wrong.
This is how I create an RSA key

```
 EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, NULL);
    if (!pctx) {
        handleErrors();
    }

    // Initialize key generation
    if (EVP_PKEY_keygen_init(pctx) <= 0) {
        handleErrors();
    }

    // Set RSA key size (2048 bits)
    if (EVP_PKEY_CTX_set_rsa_keygen_bits(pctx, 2048) <= 0) {
        handleErrors();
    }

    // Generate the private key
    EVP_PKEY* pkey = NULL;
    if (EVP_PKEY_keygen(pctx, &pkey) <= 0) {
        handleErrors();
    }
```

**Comment by ChauhanPriyanka** [comment]
I checked further EVP_PKEY_CTX_set_rsa_keygen_bits is generating 512 bits key even I'm passing 2048. Is this expected.
 int size = EVP_PKEY_get_size(pkey); // 512 is the output.

Not sure if it has some relation with this - https://github.com/openssl/openssl/issues/11863

**Reply by levitte** [reply]
`EVP_PKEY_get_size()` gets you a buffer size in bytes.  Try again with `EVP_PKEY_get_bits()`

**Reply by ChauhanPriyanka** [reply]
EVP_PKEY_get_bits() gives the correct size as - 2048. Thank you @levitte 
Still  PKCS12_set_pbmac1_pbkdf2 is giving _Error: error:02000078:rsa routines::key size too small._ What I'm missing.

**Reply by mattcaswell** [reply]
Are you sure it's that function giving you that error? I.e. could that be some stale error on the error queue from a previous failure?

**Reply by ChauhanPriyanka** [reply]
my bad, apologies for diverting the discussion, the _Error: error:02000078:rsa routines::key size too small_ is coming from some other code piece which is not relevant, thank you for pointing. 
I see that, the error inside PKCS12_set_pbmac1_pbkdf2 is initiated from here as I'm not providing salt

From p12_mult.c
```
if (salt == NULL) {
        known_salt = OPENSSL_malloc(saltlen);
        if (known_salt == NULL)
            goto err;
```

This is how I'm using
```
// Set PBKDF2 and PBMAC1 in the PKCS#12 structure
if (!PKCS12_set_pbmac1_pbkdf2(p12, password, passlen, NULL, 0, 1, md_type, prf_md_name)) {
    handleErrors();
}
```

**Comment by ChauhanPriyanka** [comment]
I added salt length  and I'm able to fix the issue. I'm able to parse the certificate also. Here is the information
```
openssl pkcs12 -info -in  output.p12 -noout
Enter Import Password:
MAC: PBMAC1 using PBKDF2, Iteration 1
Key length: 32, Salt length: 8
PBKDF2 PRF: hmacWithSHA256
PKCS7 Encrypted data: PBES2, PBKDF2, AES-128-CBC, Iteration 2048, PRF hmacWithSHA256
Certificate bag
PKCS7 Data
Shrouded Keybag: PBES2, PBKDF2, AES-128-CBC, Iteration 2048, PRF hmacWithSHA256
```
However I dont see PBMAC1 in the bag, seems ticket is open for that - https://github.com/openssl/openssl/issues/25121.

I have further question, apologies for being silly.

- Just for clarity, I will be using PKCS12_set_pbmac1_pbkdf2 in fips mode to do the mac verification, and without fips mode I will be continuing with the default mac verification using PBKDF2. Please confirm if this understanding is correct.
- I'm using different password in pkcs12_create and PKCS12_set_pbmac1_pbkdf2, so while parsing the certificate using PKCS12_parse which password I shall use,

**Reply by beldmit** [reply]
> However I dont see PBMAC1 in the bag, seems ticket is open for that - #25121.

This is not implemented yet, feel free to contribute a patch.

> Just for clarity, I will be using PKCS12_set_pbmac1_pbkdf2 in fips mode to do the mac verification, and without fips mode I will be continuing with the default mac verification using PBKDF2. Please confirm if this understanding is correct.

Sorry, I don't get. Verification is done based on the file content. Do you mean creation?

> I'm using different password in pkcs12_create and PKCS12_set_pbmac1_pbkdf2, so while parsing the certificate using PKCS12_parse which password I shall use,

Don't know, sorry

**Reply by ChauhanPriyanka** [reply]
> > Just for clarity, I will be using PKCS12_set_pbmac1_pbkdf2 in fips mode to do the mac verification, and without fips mode I will be continuing with the default mac verification using PBKDF2. Please confirm if this understanding is correct.
> 
> Sorry, I don't get. Verification is done based on the file content. Do you mean creation?
> Sorry for typo, yes MAC creation.
> > I'm using different password in pkcs12_create and PKCS12_set_pbmac1_pbkdf2, so while parsing the certificate using PKCS12_parse which password I shall use,
> 
> Don't know, sorry,
> Ok

**Comment by ChauhanPriyanka** [comment]
In our application certificate created are consumed by various devices, Android, IOS, Windows for multiple purpose.
In fips mode, certificate will use PBMAC1 along with PBKDF2, will the agents be able to read these modern certificates.
